src/main/java/org/spongepowered/common/data/processor/data/DisplayNameDataProcessor.java

```diff
@@ -40,156 +38,141 @@
-                final String displayString = mainCompound.getString(NbtDataUtil.ITEM_DISPLAY_NAME);
-                final DisplayNameData data = new SpongeDisplayNameData(SpongeTexts.fromLegacy(displayString));
-                return Optional.of(data);
+
+            final NBTTagCompound compound = ((ItemStack) holder).getSubCompound(NbtDataUtil.ITEM_DISPLAY, false);
+            if (compound != null && compound.hasKey(NbtDataUtil.ITEM_DISPLAY_NAME, 8)) {
+                return Optional.of(new SpongeDisplayNameData(SpongeTexts.fromLegacy(compound.getString(NbtDataUtil.ITEM_DISPLAY_NAME))));
             } else {
                 return Optional.empty();
             }
-        } else if (dataHolder instanceof IWorldNameable) {
-            if (((IWorldNameable) dataHolder).hasCustomName()) {
-                final String customName = ((IWorldNameable) dataHolder).getCommandSenderName();
+        } else if (holder instanceof IWorldNameable) {
+            if (((IWorldNameable) holder).hasCustomName()) {
+                final String customName = ((IWorldNameable) holder).getCommandSenderName();
                 final DisplayNameData data = new SpongeDisplayNameData(SpongeTexts.fromLegacy(customName));
                 return Optional.of(data);
+            } else {
+                return Optional.empty();
             }
         }
-        return Optional.empty();
-    }
 
-    @Override
-    public Optional<DisplayNameData> fill(DataHolder dataHolder, DisplayNameData manipulator, MergeFunction overlap) {
-        if (supports(dataHolder)) {
-            final DisplayNameData data = from(dataHolder).orElse(null);
-            final DisplayNameData newData = checkNotNull(overlap.merge(checkNotNull(manipulator), data));
-            final Text display = newData.displayName().get();
-            final boolean displays = newData.customNameVisible().get();
-            return Optional.of(manipulator.set(Keys.DISPLAY_NAME, display).set(Keys.SHOWS_DISPLAY_NAME, displays));
-        }
         return Optional.empty();
     }
 
     @Override
     public Optional<DisplayNameData> fill(DataContainer container, DisplayNameData displayNameData) {
         final String json = DataUtil.getData(container, Keys.DISPLAY_NAME, String.class);
-        final Text displayName = TextSerializers.JSON.deserialize(json);
-        final boolean shows = DataUtil.getData(container, Keys.SHOWS_DISPLAY_NAME);
-        return Optional.of(displayNameData.set(Keys.DISPLAY_NAME, displayName).set(Keys.SHOWS_DISPLAY_NAME, shows));
+        return Optional.of(displayNameData.set(Keys.DISPLAY_NAME, TextSerializers.JSON.deserialize(json)));
     }
 
     @Override
-    public DataTransactionResult set(DataHolder dataHolder, DisplayNameData manipulator, MergeFunction function) {
+    public DataTransactionResult set(DataHolder holder, DisplayNameData manipulator, MergeFunction function) {
         return DataTransactionResult.failResult(manipulator.getValues());
```

Comment:
This, and and several other methods in this class, don't need to be overriden anymore.